### PR TITLE
🚀 OOP school library: set up associations

### PR DESCRIPTION
**Pull request by jicamargo**
# 🚀 OOP school library: set up associations

📚 Project Milestone Description:
In this milestone, I've successfully implemented essential class relationships and functionalities for the School Library App. This PR showcases the following features:

1. **Classroom Implementation:**
   - Created the `Classroom` class with an instance variable `@label` initialized in the constructor.
   - Implemented setters and getters for `@label` using `attr_accessor`.

2. **Classroom-Student Relationship:**
   - Established a has-many/belongs-to relationship between `Classroom` and `Student`.
   - Implemented the has-many side (a classroom has many students).
   - Implemented the belongs-to side (a student belongs to a classroom).
   - Ensured that adding a student to a classroom also sets the classroom for the student.
   - Ensured that setting the classroom for a student also adds the student to the classroom's students.

3. **Book Implementation:**
   - Created the `Book` class with instance variables `@title` and `@author` initialized in the constructor.
   - Implemented setters and getters for instance variables using `attr_accessor`.

4. **Rental Implementation:**
   - Created the `Rental` class with an instance variable `@date` initialized in the constructor.
   - Implemented setters and getters for `@date` using `attr_accessor`.

5. **Person-Book Relationship (Many-to-Many):**
   - Established a many-to-many relationship between `Person` and `Book` using the intermediate class `Rental`.
   - Implemented the has-many side of `Book` and `Rental` (a book has many rentals).
   - Implemented the belongs-to side of `Rental` and `Book` (a rental belongs to a book).
   - Implemented the has-many side of `Person` and `Rental` (a person has many rentals).
   - Implemented the belongs-to side of `Rental` and `Person` (a rental belongs to a person).
   - Modified the constructor of `Rental` to set `Book` and `Person`.

👍 Please review these changes to move forward with the project development. Your feedback and collaboration are highly appreciated!

### CODE REVIEW

File: book.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+# Purpose: Book class
+#
+class Book
+  attr_accessor :title, :author
+
+  def initialize(title, author)
+    @title = title
+    @author = author
+    @rentals = []
+  end
```
Comment (mikethreels) [inline book.rb]: - [x] to make the many to many relationship work Getters and Setters for @rentals should be present
- [x] A method for adding rentals should also be present. please see this [Link](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#many-to-many-relationship) to the example in a previous lesson.

File: rental.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+    # many to many relationship between book and person, using a join class (rental)
+    #
+    # setter for book (a rental can only belong to one book)
+    @book = book
+    book.rentals << self
+
+    # setter for person (a rental can only belong to one person)
+    @person = person
+    person.rentals << self
```
Comment (mikethreels) [inline rental.rb]: - [x] to make this relationship work Getters and Setters for @book and @person instance variables should be present

File: person.rb [hunk 3]
```diff
@@ -1,26 +1,21 @@
...
-end
+require_relative 'nameable'
 
 #
 # Class representing a person in the library application
 #
 class Person < Nameable
   attr_accessor :name, :age
   attr_writer :id
+  attr_reader :rentals
```
Comment (mikethreels) [inline person.rb:11]: - [x] to make this relationship work Getters and Setters for @rentals should be present

File: person.rb [hunk 4]
```diff
@@ -1,26 +1,21 @@
...
 
   def initialize(age, name = 'Unknown', parent_permission: true)
     super()
     @name = name
     @age = age
     @parent_permission = parent_permission
+    @rentals = []
```
Comment (mikethreels) [inline person.rb:19]: - [x] A method for adding rentals should also be present. please see this [Link](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#many-to-many-relationship) to the example in a previous lesson.

**Review by mikethreels — CHANGES_REQUESTED** [review]
Hi Jorge, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
- good job setting up the associations.

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: book.rb (hunk 1 above)
Comment (jicamargo) [reply book.rb]: Thank you. changes done!

File: rental.rb (hunk 2 above)
Comment (jicamargo) [reply rental.rb]: Thank you. changes Done!

File: person.rb (hunk 3 above)
Comment (jicamargo) [reply person.rb:11]: Thank you. changes done!

File: person.rb (hunk 4 above)
Comment (jicamargo) [reply person.rb:19]: changes done, thank you!

**Review by brainconnect93 — APPROVED** [review]
## Status: Approved :heavy_check_mark:

Hi @jicamargo, 👋 👋  

Your project is complete! Great job for following the project requirement🥇and implementing the changes requested by the previous reviewer. There is nothing else to say other than... it's time to merge it 🍾🚢 :shipit: . Congratulations!  🎉💯🌟 




## Good Points 👍

- Descriptive Pull Request title & summary. ✔️
- set up associations for school library✔️
- Good use of Git flow ✔️

## Optional changes

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers..Happy coding!..and keep soaring higher! 💻 🍷 🚀 

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @brainconnect93 in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have 2 more limited reviews per this project. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request closed without merging**